src/baal/active/heuristics/heuristics.py

```diff
@@ -435,3 +443,111 @@ def __init__(self, shuffle_prop=0.0, threshold=None, reverse=False):
+    if the relevant model output to that heuristic has more than 3-dimenstions.
+    In `CombineHeuristics`, the `reduction` is used to aggregate the final result of
+    heuristics.
+
+    Args:
+        heuristics (list[AbstractHeuristic]): list of heuristic instances
+        weights (list[float]): the assigned weights to the result of each heuristic
+            before calculation of ranks
+        reduction (Union[str, callable]): function that aggregates the results of the heuristics
+            (default: weighted average which could be used as (reduction='mean`)
+        **kwargs : The only used keyword is `shuffle_prop` similar usage as `AbstractHeuristic`
+
+
+    """
+    def __init__(self, heuristics: list(), weights: list(), reduction='mean', **kwargs):
+        super(CombineHeuristics, self).__init__(reduction=reduction, **kwargs)
+        self.composed_heuristic = heuristics
+        self.weights = weights
+
+        reversed = [bool(heuristic.reversed) for heuristic in self.composed_heuristic]
+
+        if all(item is False for item in reversed):
+            self.reversed = False
+        elif all(item is True for item in reversed):
+            self.reversed = True
+        else:
+            raise Exception("heuristics should have the same value for `revesed` parameter")
+
+        self.threshold = [bool(heuristic.threshold) for heuristic in self.composed_heuristic]
+
+    def get_uncertainties(self, predictions):
+        """
+        Computes the score for each part of predictions according to the assigned heuristic.
+
+        NOTE: predictions is a list of each model outputs. For example for a object detection model,
+        the predictions should be as:
+            [confidence_predictions: nd.array(), boundingbox_predictions: nd.array()]
+
+        Args:
+            predictions (list[ndarray]): list of predictions arrays
+
+        Returns:
+            Array of uncertainties
+
+        """
+
+        results = []
+        for ind, prediction in enumerate(predictions):
+            if isinstance(predictions[0], types.GeneratorType):
+                results.append(self.composed_heuristic[ind].get_uncertainties_generator(prediction))
+            else:
+                results.append(self.composed_heuristic[ind].get_uncertainties(prediction))
+        return results
+
+    def get_ranks(self, predictions):
+        """
+        Rank the predictions according to the weighted vote of each heuristic.
+
+        Args:
+            predictions (list[ndarray]):
+                list[[batch_size, C, ..., Iterations], [batch_size, C, ..., Iterations], ...]
+
+        Returns:
+            Ranked index according to the uncertainty (highest to lowest).
+
+        """
+
+        scores_list = self.get_uncertainties(predictions)
+
+        # normalizing weights
+        w = np.array(self.weights).sum()
+        self.weights = [weight / w for weight in self.weights]
```

Comment:
Should we normalize it? Does it matter?

Reply:
I don't know if it matters... but to give a respective weight. because at the end we want to make a decision based on the result of each of the heuristics, so having a relative weight is always better I assume?